backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java

```diff
@@ -0,0 +1,149 @@
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.UPDATE)
+        .thenCompose(unused -> attributeHandler.deleteKeyValues(projectId, maybeKeys), executor)
+        .thenCompose(unused -> updateModifiedTimestamp(projectId, now), executor);
+  }
+
+  public InternalFuture<List<KeyValue>> getAttributes(GetAttributes request) {
+    final var projectId = request.getId();
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.READ)
+        .thenCompose(unused -> attributeHandler.getKeyValues(projectId), executor);
+  }
+
+  public InternalFuture<Void> logAttributes(LogAttributes request) {
+    final var projectId = request.getId();
+    final var attributes = request.getAttributesList();
+    final var now = Calendar.getInstance().getTimeInMillis();
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.UPDATE)
+        .thenCompose(unused -> attributeHandler.logKeyValues(projectId, attributes), executor)
+        .thenCompose(unused -> updateModifiedTimestamp(projectId, now), executor);
+  }
+
+  public InternalFuture<Void> addTags(AddExperimentRunTags request) {
+    final var projectId = request.getId();
+    final var tags = request.getTagsList();
+    final var now = Calendar.getInstance().getTimeInMillis();
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.UPDATE)
+        .thenCompose(unused -> tagsHandler.addTags(projectId, tags), executor)
+        .thenCompose(unused -> updateModifiedTimestamp(projectId, now), executor);
+  }
+
+  public InternalFuture<Void> deleteTags(DeleteExperimentRunTags request) {
+    final var projectId = request.getId();
+    final var now = Calendar.getInstance().getTimeInMillis();
+
+    final Optional<List<String>> maybeTags =
+        request.getDeleteAll() ? Optional.empty() : Optional.of(request.getTagsList());
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.UPDATE)
+        .thenCompose(unused -> tagsHandler.deleteTags(projectId, maybeTags), executor)
+        .thenCompose(unused -> updateModifiedTimestamp(projectId, now), executor);
+  }
+
+  public InternalFuture<List<String>> getTags(GetTags request) {
+    final var projectId = request.getId();
+
+    return checkProjectPermission(projectId, ModelDBActionEnum.ModelDBServiceActions.READ)
+        .thenCompose(unused -> tagsHandler.getTags(projectId), executor);
+  }
+
+  private InternalFuture<Void> updateModifiedTimestamp(String projectId, Long now) {
+    return jdbi.useHandle(
+        handle ->
+            handle
+                .createUpdate(
+                    "update project set date_updated=greatest(date_updated, :now) where id=:project_id")
+                .bind("project_id", projectId)
+                .bind("now", now)
+                .execute());
+  }
+
+  private InternalFuture<Void> checkProjectPermission(
```

Comment:
Probably should be public so that other things (e.g. ER DAO) can call instead of duplicating.